lib/psych/coder.rb

```diff
@@ -66,6 +66,7 @@ def represent_object tag, obj
     # Emit a scalar with +value+
     def scalar= value
       @type   = :scalar
+      @style  = Psych::Nodes::Scalar::ANY if @style == Psych::Nodes::Mapping::BLOCK
```

Comment:
Is this actually necessary?  If I remove this line, all of the tests including the ones you added still pass.

Reply:
It should be. I'm not completely sure, but I suspect that there are two issues which prevents the code to fail without it.
1) In ext/psych/yaml/emitter.c the YAML_ANY_SCALAR_STYLE is handled exactly same as YAML_PLAIN_SCALAR_STYLE - you just can't force plain style on something which would be invalid with it. Which can hardly be called bug, of course.
2) Psych::Nodes::Scalar::PLAIN is numerically equal to Psych::Nodes::Mapping::BLOCK - which is not bug, but can be seen as arbitrary and just lucky.

It will become very hard to find bug if the condition 2 ever changes, but I'm not sure how to test for it. Condition 1 MIGHT be assumed to be permanent.

Reply:
Actually, the code will also break if Psych::Nodes::Mapping::BLOCK and Psych::Nodes::Sequence::BLOCK will ever differ. I mean Psych, not just my changes.

Reply:
Can you write a test that fails if I remove that line?

Reply:
Not without changing the C part to work worse. However, If you really want to remove that line, I can make test which fails when Psych::Nodes::Mapping::BLOCK, Psych::Nodes::Sequence::BLOCK and Psych::Nodes::Scalar::PLAIN aren't numerically equal, with comments about why that's important (basically, copying the explanation from here).

Or, maybe the line "@style    = Psych::Nodes::Mapping::BLOCK" in Psych::Coder::initialize should be commented with note that without such numerical equivalence the code will break.